broadcastclients/broadcastclients.go

```diff
@@ -72,12 +119,52 @@ func (bcs *BroadcastClients) adjustCount(delta int32) {
 }
 
 func (bcs *BroadcastClients) Start(ctx context.Context) {
-	for _, client := range bcs.clients {
+	bcs.router.StopWaiter.Start(ctx, bcs.router)
+
+	for _, client := range bcs.primaryClients {
 		client.Start(ctx)
 	}
+
+	bcs.router.LaunchThread(func(ctx context.Context) {
+		startNewFeedTimer := time.NewTicker(MAX_FEED_INACTIVE_TIME)
+		defer startNewFeedTimer.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case cs := <-bcs.router.confirmedSequenceNumberChan:
+				startNewFeedTimer.Stop()
+				bcs.router.forwardConfirmationChan <- cs
+				startNewFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
+			case msg := <-bcs.router.messageChan:
+				startNewFeedTimer.Stop()
+				if err := bcs.router.forwardTxStreamer.AddBroadcastMessages([]*broadcaster.BroadcastFeedMessage{&msg}); err != nil {
+					log.Error("Error routing message from Sequencer Feed", "err", err)
+				}
+				startNewFeedTimer.Reset(MAX_FEED_INACTIVE_TIME)
+			case <-startNewFeedTimer.C:
+				// failed to get messages from primary feed for ~5 seconds, start a new feed
+				bcs.StartSecondaryFeed(ctx)
```

Comment:
Paraphrasing how this code works to organize my thoughts: Whenever there's 6 seconds of no messages then the secondaries will get started, in the order provided in the list. Secondaries will never be stopped once starting. For chains where there is less activity then the secondaries will all be started. Deduping of messages is the responsibility of downstream `forwardTxStreamer`/`forwardConfirmationChan`, it is okay to have many clients writing the same message to the channel.

Thinking about this now, it seems a little weird to leave the secondaries on forever if they are ever switched to. Why not just make them primaries from the beginning in that case? If the concern is about having too many clients running (maybe bandwidth concerns?), if we never turn off the secondaries then we can end up in that situation anyway of having too many clients running.

Reply:
Initially I had set the timer to 6 seconds to handle cases for the delays incurred when starting a new feed, I'll change it back to 5 seconds since we are stopping and restarting the timer.

handling of deduping of messages and confirmations inside router makes more sense since this has added benefits regarding usage of broadcastclients on the node side too.

regarding leaving the secondaries on forever, do you think we can utilize deduping logic to figure out the minimum of repetitions over a set of message and if that exceeds a threshold we could go ahead and stop the latest started secondary?